_posts/2020-06-18-Metal3-dev-env-BareMetal-Cluster-Deployment.md

```diff
@@ -0,0 +1,396 @@
+### Provision Workers
+
+The script is located at `${metal3-dev-env-path}/scripts/v1alphaX/provision_worker.sh` and it provisions a node to be added as a worker to the baremetal cluster. It selects one of the remaining nodes and provisions it and adds it to the baremetal cluster ( which only has a master at this point ). The resources created for workers are - `MachineDeployment` which can be scaled up to add more workers to the cluster, and it also created a `MachineSet` which then creates a `Machine` managing the node.
+
+> info "Note"
+> Similar to a Controlplane provisioning, a Worker provisioning also takes some time, and you can watch the process using steps shared a bit later. This will also apply when you scale Up/Down workers at a later point in time.
+
+<br/>
+This is what a `MachineDeployment` looks like
+
+```console
+kubectl describe MachineDeployment ${CLUSTER_NAME} -n metal3
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachineDeployment
+metadata:
+  [....]
+  ownerReferences:
+  - apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: bmetalcluster
+    uid: aec0f73b-a068-4992-840d-6330bf943d22
+  resourceVersion: "66257"
+  selfLink: /apis/cluster.x-k8s.io/v1alpha3/namespaces/metal3/machinedeployments/bmetalcluster
+  uid: f598da43-0afe-44e4-b793-cd5244c13f4e
+spec:
+  clusterName: bmetalcluster
+  minReadySeconds: 0
+  progressDeadlineSeconds: 600
+  replicas: 1
+  revisionHistoryLimit: 1
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: bmetalcluster
+      nodepool: nodepool-0
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 0
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: bmetalcluster
+        nodepool: nodepool-0
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+          name: bmetalcluster-workers
+      clusterName: bmetalcluster
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: Metal3MachineTemplate
+        name: bmetalcluster-workers
+      version: v1.18.0
+status:
+  observedGeneration: 1
+  phase: ScalingUp
+  replicas: 1
+  selector: cluster.x-k8s.io/cluster-name=bmetalcluster,nodepool=nodepool-0
+  unavailableReplicas: 1
+  updatedReplicas: 1
+```
+
+<br/>
+To check the status we can follow steps similar to Controlplane case :
+
+```console
+kubectl get bmh -n metal3 -w
+```
+> We can see the live status of the node being provisioned. As mentioned before `bmh` is short representation for `BareMetalHosts`.
+
+```console
+kubectl get Machine -n metal3 -w
+```
+> This shows the status of Machines associated with workers, apart from the one for Controlplane, and we can watch the status of provisioning under PHASE 
+
+```console
+sudo virsh net-dhcp-leases baremetal
+```
+> To get the nodes IP
+
+```console
+ssh metal3@{master-node-ip}
```

Comment:
you migth want to do a kubectl get nodes on the master to check that

Reply:
Done.